CMake: Allow overriding linker

Sonicadvance1 opened the pull request:
CMake: Allow overriding linker

While the ENABLE_LLD and ENABLE_MOLD options are nice, they don't handle
the case when the linker of `lld` or `mold` doesn't match the compiler.

This particularly crops up when overriding the C compiler to a new
version of clang but the globally installed `ld.lld` is still the old
clang version.
This then causes clang to fail with unusual errors when upstream breaks
compatibility with itself.

Easy enough to use by passing the linker to cmake:
`-DUSE_LINKER=/usr/bin/ld.lld-15`

This also removes the ENABLE_LLD and ENABLE_MOLD options to use
USE_LINKER directly.
- ldd: `-DUSE_LINKER=lld`
- mold: `-DUSE_LINKER=mold`

Example of compiler failure when built with clang-15 but attempting to
link with ld.lld 14:
```bash
ld.lld-14: error: unittests/APITests/CMakeFiles/Filesystem.dir/Filesystem.cpp.o: Opaque pointers are only supported in -opaque-pointers mode (Producer: 'LLVM15.0.7' Reader: 'LLVM 14.0.6')
```

neobrain commented:
Can we drop `ENABLE_LLD` and `ENABLE_MOLD` in favor of this new setting? IIRC the only reason for adding `ENABLE_MOLD` as a separate variable was that `LD_OVERRIDE` triggered special behavior elsewhere in `CMakeLists.txt`, but that is not the case anymore.

We can also use a simpler variable name then (e.g. `USE_LINKER`), but no strong opinions on that.

Sonicadvance1 commented:
> Can we drop `ENABLE_LLD` and `ENABLE_MOLD` in favor of this new setting? IIRC the only reason for adding `ENABLE_MOLD` as a separate variable was that `LD_OVERRIDE` triggered special behavior elsewhere in `CMakeLists.txt`, but that is not the case anymore.
> 
> We can also use a simpler variable name then (e.g. `USE_LINKER`), but no strong opinions on that.

Removed ENABLE_LLD and ENABLE_MOLD. Switched over to USE_LINKER name.

neobrain commented on CMakeLists.txt:
```diff
@@ -157,13 +156,9 @@ endif()
-  add_link_options(${LD_OVERRIDE})
-elseif (ENABLE_MOLD)
-  add_link_options("-fuse-ld=mold")
+if (NOT USE_LINKER STREQUAL "")
```
Empty strings evaluate to false already:

```suggestion
if (USE_LINKER)
```

Sonicadvance1 replied:
resolved.
